cou/commands.py

```diff
@@ -0,0 +1,302 @@
+        usage="cou plan data-plane [options]",
+        parents=[subcommand_common_opts_parser, dp_subparser],
+        formatter_class=CapitalizeHelpFormatter,
+    )
+
+    # Arg parser for "cou run" sub-command and set up common options
+    run_args_parser = argparse.ArgumentParser(add_help=False)
+    run_args_parser.add_argument(
+        "--interactive",
+        help="Run upgrade with prompt.",
+        action=argparse.BooleanOptionalAction,
+        default=True,
+    )
+    run_parser = subparsers.add_parser(
+        "run",
+        description="Run the cloud upgrade. If upgrade-group is unspecified, perform "
+        "upgrade for the entire cloud",
+        help="Run the cloud upgrade.",
+        usage="cou run [options]",
+        parents=[subcommand_common_opts_parser, run_args_parser],
+        formatter_class=CapitalizeHelpFormatter,
+    )
+
+    # Create control-plane and data-plane child commands to run partial upgrades
+    run_subparser = run_parser.add_subparsers(
+        title="Upgrade group",
+        dest="upgrade-group",
+        help="For more information about an upgrade group, run 'cou run <upgrade-group> -h'.",
+    )
+    run_subparser.add_parser(
+        "control-plane",
+        description="Run upgrade for the control-plane components.",
+        help="Run upgrade for the control-plane components.",
+        usage="cou plan control-plane [options]",
+        parents=[subcommand_common_opts_parser, run_args_parser],
+        formatter_class=CapitalizeHelpFormatter,
+    )
+    run_subparser.add_parser(
+        "data-plane",
+        description="Run upgrade for the data-plane components. This is possible only if "
+        "control-plane has been fully upgrade. Otherwise an error will be thrown.",
+        help="Run upgrade for the data-plane components. This is possible only if "
+        "control-plane has been fully upgrade. Otherwise an error will be thrown.",
+        usage="cou plan data-plane [options]",
+        parents=[
+            subcommand_common_opts_parser,
+            dp_subparser,
+            run_args_parser,
+        ],
+        formatter_class=CapitalizeHelpFormatter,
+    )
+
+    # It no sub-commands or options are given, print help message and exit
+    if len(sys.argv[1:]) == 0:
+        parser.print_help()
+        sys.exit(0)
+
+    try:
+        parsed_args = parser.parse_args(args)
+
+        # print help messages for an available sub-command
+        if parsed_args.command == "help":
+            match parsed_args.subcommand:
+                case "run":
+                    run_parser.print_help()
+                case "plan":
+                    plan_parser.print_help()
+                case "all":
+                    parser.print_help()
+            sys.exit(0)
+
+        return parsed_args
+    except argparse.ArgumentError as exc:
+        print(f"Error parsing arguments: {exc}\n")
+        print("See 'cou help' for more information.")
+        sys.exit(1)
```

Comment:
To me this should not be part of parse_args function. We are doing CLI logic here + sys.exit.

Reply:
I agree that this fits best in entrypoint function, but since it need to access subparsers to print help messages, I implemented it on this level. 

A different solution would be let parse_args function, instead of returning parsed args, return both parser and subparser to entrypont, which I have implemented in the new patch. Please let me know if this is better.

Reply:
Thanks for the suggestion. I have moved the parsing exception handling back to `parse_args` function and uses `parser.error`/'parser.exit' instead.